pkg/stanza/fileconsumer/internal/reader/compressedreader.go

```diff
@@ -0,0 +1,147 @@
+package reader
+
+import (
+	"bytes"
+	"compress/gzip"
+	"errors"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/header"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/scanner"
+	"go.uber.org/zap"
+	"golang.org/x/net/context"
+	"math"
+	"os"
+)
+
+type CompressedReader struct {
+	*Reader
+
+	buffer           *bytes.Buffer
+	lastReadFileSize int64
+	fromBeginning    bool
+}
+
+func (r *CompressedReader) ReadToEnd(ctx context.Context) {
+	newFileHandle, err := os.Open(r.file.Name())
+	if err != nil {
+		r.set.Logger.Error("problem creating reader for compressed file", zap.Error(err))
+		return
+	}
+	r.file = newFileHandle
+	gzipReader, err := gzip.NewReader(r.file)
+	if err != nil {
+		r.set.Logger.Error("problem creating reader for compressed file", zap.Error(err))
+		return
+	}
+
+	// TODO validate if we can keep this check for the file size to avoid unnecessary decompression
+	// in theory, this check can lead to new log entries going unnoticed, if the new entries are duplicates
+	// of a previous entry, as this leads to the size of the compressed file not being increased.
+	// practically this may not be an issue for logs, since at least the timestamp should change for a new log entry,
+	// even if it has the same message as the line before
+
+	stat, err := r.file.Stat()
+	if err != nil {
+		r.set.Logger.Error("problem creating determining size of compressed file", zap.String("fileName", r.fileName), zap.Error(err))
+		return
+	}
+
+	initalRead := true
+	if lastReadFileSize, ok := r.FileAttributes["lastReadFileSize"]; ok {
```

Comment:
I'm open to adding such an attribute but it should be proposed in another issue.

In another comment I question whether we need to track this for the sake of this implementation. If we don't, let's remove it. If we do, let's use an unexported field on the reader struct.